A string to control the layout of the dlg_features in log messages

Priyeshkkumar opened the pull request:
The current version of dlg prints the dlg_features in a specific order i.e. **[time file_line function_name {tags}]**
This pull_request introduces a variable **dlg_log_format** of type **const char*** through which the user can change the order in which these dlg_features are printed.

Priyeshkkumar committed:
Add a string that controls how logs are printed

nyorain commented:
I'm not convinced this is the best way to solve this. I definitely see your usecase (`dlg_generic_output` is quite limited) but there is absolutely no need to call it in your dlg handler. You could just print the output in the format you want/need. I don't even expect `dlg_generic_output` to be useful for many (especially more complex) applications, as they might want to have custom output. If every user just modifies and extends `dlg_generic_output` to fit their needs, this will make maintaining dlg a total mess. It's mainly there as a simple fallback and you seem to have quite some requirements regarding your output so you should probably just write the custom printf calls you want (or log to a file or whatever).
If you have any reason for not doing that, let me know! I'm not completely opposed to merging something like this but it would need a considerable amount of work to begin with, I see some issues with the current implementation.

Priyeshkkumar commented:
> I'm not convinced this is the best way to solve this. I definitely see your usecase (`dlg_generic_output` is quite limited) but there is absolutely no need to call it in your dlg handler. You could just print the output in the format you want/need. I don't even expect `dlg_generic_output` to be useful for many (especially more complex) applications, as they might want to have custom output. If every user just modifies and extends `dlg_generic_output` to fit their needs, this will make maintaining dlg a total mess. 

I understand your concerns but consider a scenario where a user wants to keep using the **dlg_features** with **dlg_generic_output()** and just wants to change the layout of the additional information in the logs, from the default layout i.e. **[time file_line function_name {tags}]** to something else maybe tags first and then the rest of the additional information, with the current implementation of dlg the user will have to completely write the code to handle the layout + code for printing time-stamp + code for thread-saftey, which I think would be an overhead for the user who just wants to change the layout of the additional information in the logs.

>It's mainly there as a simple fallback and you seem to have quite some requirements regarding your output so you should probably just write the custom printf calls you want (or log to a file or whatever).

I like the way how dlg provides the **dlg_features**  along with the **dlg_generic_output()** ,it makes dlg pretty much customizable for users who do not want to use there own log handler.
Therefore, I was thinking why not make it more customizable...

> If you have any reason for not doing that, let me know! I'm not completely opposed to merging something like this but it would need a considerable amount of work to begin with, I see some issues with the current implementation.

I have updated the PR to solve the issues, but if you find any more concerns regarding the implementation we could discuss and change it accordingly. But it would be nice to have a feature like this to customize the additional information in a log message for users who do not want to use there own log handler...

nyorain commented:
I will consider this some more. If it does not end up in dlg, I would at least like to put it to the docs, if you don't mind.
Regarding the problems:
- I don't like he way `features` and `dlg_log_format` are currently combined, it makes for a very confusing interface that you have to specify something in features *and* in the format string to be actually printed.
- please add a new function instead of changing the exsiting one. I pretty much try to keep dlg backwards compatible (the library is so small in scope that it's useful) and in this case, we could simply add a new function, something like `dlg_generic_output_formatted`. This also solves the problem above: the new function would not need the `features` parameter anymore and from `dlg_generic_output` (and related functions) we could simply call the new function with a format string built from the features.
- tags isn't correctly comma separated anymore (in fact, it seems like tags aren't separated at all?)
- the handling of the format string is confusing, users would probably expect it to function printf-like but it doesn't. The '%' char can always be ommitted and invalid chars will simply be skipped (for printf-like behavior they would have to be outputted, that makes more sense IMO).
- minor nitpick: the codestyle does not match what the rect of the library uses in several places
- the vscode file does not belong into the repo, feel free to add it to gitignore though

Priyeshkkumar commented:
> * I don't like he way `features` and `dlg_log_format` are currently combined, it makes for a very confusing interface that you have to specify something in features _and_ in the format string to be actually printed.
> * please add a new function instead of changing the exsiting one. I pretty much try to keep dlg backwards compatible (the library is so small in scope that it's useful) and in this case, we could simply add a new function, something like `dlg_generic_output_formatted`. This also solves the problem above: the new function would not need the `features` parameter anymore and from `dlg_generic_output` (and related functions) we could simply call the new function with a format string built from the features.

It is actually a great idea...

> * tags isn't correctly comma separated anymore (in fact, it seems like tags aren't separated at all?)

I will correct it in the next commit...

> * the handling of the format string is confusing, users would probably expect it to function printf-like but it doesn't. The '%' char can always be ommitted and invalid chars will simply be skipped (for printf-like behavior they would have to be outputted, that makes more sense IMO).

I think I am not able to understand your point clearly here, could you provide me an example?
If you are thinking of making the new function to be a printf like format, something like:
`dlg_generic_output_formatted( "[ line:%d and file:%s ]", *origin->line, *origin->file);`
then how are you planning the support of `dlg_features` like `dlg_output_threadsafe`, and `dlg_output_time` or `dlg_output_time_msecs` as these are not available in  `struct dlg_origin`   ?
 
> * minor nitpick: the codestyle does not match what the rect of the library uses in several places

I will try to be more considerate about it from now on.

Priyeshkkumar commented:
Any Updates on this?

nyorain commented:
What I meant regarding the new formatting: The current implementation handles some cases in a confusing way that I would not consider expected behavior. The interface is very `printf`-like and users will likely expect it to behave somewhat like it. For instance:
- `dlg_generic_output("test", ...)` would just output the time in seconds, twice. I would rather expect it to just output "test". This means: really check that only chars after '%' are treated as format specifiers and pass through all other chars (the way you currently pass through '(', '[', '{' and so on, why should they be special?).
- Invalid format specifiers are not handled at all. I'm not sure how this should be handled best, but just ignoring it is not a good idea. Maybe replace them by `<invalid format specifier>` or something?

Users would currently have to learn a whole set of new format specifiers to use this function, this makes the interface hard to use. To be honest, seeing how many parts are here to get wrong or that might confuse users, I would much rather have this as a separate contribution to the docs that users can use in their codebase if they need it instead of integrating this into dlg. It's just out of scope of the main library, even if it's useful.

Priyeshkkumar commented:
As suggested by you I have removed the previous implementation and added three new functions which will do the desired work 
and will keep dlg backward compatible.
* I have added a new function `dlg_generic_output_formatted()` which will print the `dlg_features` according to the string passed to it. I have also provided a list of all conversion characters which will represent the `dlg_features`, in file **include/dlg/dlg.h:240**, only these conversion characters will be converted to respective `dlg_features`, rest of the characters will be printed as it is. 
* I have also added a setter function `dlg_set_layout( const char* )` which will set a custom layout of dlg_features printed along with the actual message, and it is completely up to the user to use or to not use this function. 
  * If the user uses this function to set the custom layout, and then afterwards uses any of the `dlg_generic_output_stream()` or `dlg_generic_output_buf()` function then the program flow will automatically go to the new function `dlg_generic_output_formatted()` which will print the `dlg_features` according to the layout passed to `dlg_set_layout()` function.
I have also updated the **docs/tests/core.c** file to showcase these changes. 
  * If the user does not use this function then there will be no effect on dlg. 

* I have also provided a function `dlg_set_default_layout()`, it takes no arguments and it resets the default layout.

Priyeshkkumar commented:
Any Updates?

nyorain commented:
Sorry for the delay. The formatting function looks better now. Adding a new global seems like an overkill, I'd prefer we don't do this. The way you have to use "features" and the format string is still confusing. When I find the time I can fix those things up myself though and merge it.

Priyeshkkumar commented:
> Sorry for the delay. The formatting function looks better now. Adding a new global seems like an overkill, I'd prefer we don't do this. The way you have to use "features" and the format string is still confusing. When I find the time I can fix those things up myself though and merge it.

Thanks, I would be very grateful :-)

nyorain committed:
Rework the formatted output feature

nyorain commented:
Ok so I squashed your commits and fixed the issues I saw with it. Namely:
- don't introduce a new global
- cleanly separate the format string from the dlg features
- some minor issues regarding the format specifiers and the formatting function, lot of codestyle changes.
Let me know if it works for your purpose like this, I'd consider this ready for merging if CI passes.

nyorain committed:
Fix warning/error on msvc

Priyeshkkumar commented:
> Ok so I squashed your commits and fixed the issues I saw with it. Namely:
> 
> * don't introduce a new global
> * cleanly separate the format string from the dlg features
> * some minor issues regarding the format specifiers and the formatting function, lot of codestyle changes.

Thanks 😊

>   Let me know if it works for your purpose like this, I'd consider this ready for merging if CI passes.

It is working stupendously🎉🎉
Thanks
